content/docs/concepts/remote_write_spec_2_0.md

```diff
@@ -0,0 +1,394 @@
+  // text for the metric.
+  uint32 help_ref = 3;
+  // unit_ref is a reference to the Request.symbols array representing unit
+  // for the metric.
+  uint32 unit_ref = 4;
+}
+
+// A native histogram, also known as a sparse histogram.
+message Histogram { ... }
+
+// A BucketSpan defines a number of consecutive buckets with their
+// offset.
+message BucketSpan { ... }
+```
+
+All timestamps MUST be int64 counted as milliseconds since the Unix epoch. Sample's values MUST be float64.
+
+For every `TimeSeries` message:
+
+* Label references MUST be provided.
+* At least one element in Samples or in Histograms MUST be provided. For series which (rarely) would mix float and histogram samples, a separate `TimeSeries` message MUST be used.
+<!---
+TODO(bwplotka): We have some inconsistency here. In gdoc we talked this should be SHOULD. But proto has it as MUST (?). What do we want at the end? What's wrong with MUST here?
+-->
+* Metadata MUST be provided.
+* Exemplars SHOULD be provided, if they exist.
+* Created timestamp SHOULD be provided for metrics that follow counter semantics (e.g. counters and histograms).
+
+The following subsections define some schema elements in details.
+
+#### Symbols
+
+The `io.prometheus.write.v2.Request` proto schema is designed to [intern all strings](https://en.wikipedia.org/wiki/String_interning) for the proven additional compression and memory efficiency gains on top of the standard compressions.
+
+Symbols table containing deduplicated strings used in series and exemplar labels, metadata strings MUST be provided. References MUST point to the existing index in the Symbols string array.
+
+#### Series Labels
+
+The complete set of labels MUST be sent with each Sample or Histogram sample. Additionally, the label set associated with samples:
+
+- SHOULD contain a `__name__` label.
+- MUST NOT contain repeated label names.
+- MUST have label names sorted in lexicographical order.
+- MUST NOT contain any empty label names or values.
+
+Sender MUST only send valid metric names, label names, and label values:
+
+<!---
+TODO(bwplotka): Add mention for UTF-8 support going forward, see https://docs.google.com/document/d/1PljkX3YLLT-4f7MqrLt7XCVPG3IsjRREzYrUzBxCPV0/edit?disco=AAAA4gSqQ7g
+-->
+- Metric names MUST adhere to the regex `[a-zA-Z_:]([a-zA-Z0-9_:])*`.
+- Label names MUST adhere to the regex `[a-zA-Z_]([a-zA-Z0-9_])*`.
+- Label values MAY be any sequence of UTF-8 characters .
+
+Receiver MAY impose limits on the number and length of labels, but this will be receiver-specific and is out of scope for this document.
+
+Label names beginning with "__" are RESERVED for system usage and SHOULD NOT be used, see [Prometheus Data Model](https://prometheus.io/docs/concepts/data_model/).
+
+#### Samples
+
+<!---
+TODO(bwplotka): Does this apply to histograms?
+-->
+Sender MUST send samples for any given TimeSeries in timestamp order. Sender MAY send multiple requests for different series in parallel.
```

Comment:
Technically prometheus supports out of order samples now, and I think by proxy remote write ingestion via the write handler does as well? Should this be `SHOULD`? or `MUST, unless via some external system an out of order sample is generated such that a sample has a timestamp older than a previously sent sample`?